[Compiler] Design Draft

zhoonit opened the pull request:
This patch defines interface of the graph compiler we've talked through.

I changed the design little bit that graph rep has interpreter and compiler.
The design proposed in this PR is ready-to-wrap to the discussed design.

- [Compiler/skeleton] Set up compiler source files
```
This patch setup compiler source files

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>
```
- [Compiler] Add concept of the compiler
```
This patch adds a concept of the compiler and class declaration.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>
```
- [Compiler] Declare interface
```
This patch declares interface of compiler / interpreter

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>
```
- [Compiler] Add Graph Representation
```
This patch adds graph representation and it's interface

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>
```

jijoongmoon commented:
I recommend to draw class diagram or UML.

kparichay commented on nntrainer/compiler/compiler.h:
```diff
@@ -0,0 +1,209 @@
+ * @date 01 April 2021
+ * @brief NNTrainer compiler that reads and generates executable graph
+ * @details
+ * Graph is convertible either to iostream, representation, executable by
```
add some scenarios of iostream like tflite format, ini, etc.

zhoonit replied:
Updated

kparichay commented:
As the model class is going to interact with ExecutableGraph, it would be good to add some interface for this as well.
Note that you can just borrow the NetworkGraph interface for now to just support existing functionality and extend later.
Also, you might want to list out the differences in the GraphRepresentation and ExecutableGraph.

Further, add the invariants the GraphRepresentation - that it is the most expanded version of the graph (where the most expanded version can change depending on the granularity with which nntrainer works). Also keep in mind to keep the support required by c/c++ API for creating and adding layers or certain optimizations which depends on the graph.

zhoonit commented:
> I recommend to draw class diagram or UML to see whole picture.

Class diagram has been updated on the confluence :)

zhoonit commented:
> As the model class is going to interact with ExecutableGraph, it would be good to add some interface for this as well.
> Note that you can just borrow the NetworkGraph interface for now to just support existing functionality and extend later.

I'll do this with a new pr. I am thinking to copy the interface as much as possible to not affect the interaction and to not invent the wheel again.

> Also, you might want to list out the differences in the GraphRepresentation and ExecutableGraph.

updated.

> Further, add the invariants the GraphRepresentation - that it is the most expanded version of the graph (where the most expanded version can change depending on the granularity with which nntrainer works). Also keep in mind to keep the support required by c/c++ API for creating and adding layers or certain optimizations which depends on the graph.

Added as a note.

zhoonit commented:
v2. 

- Reverted GraphRepresentation / ExecutableGraph to alias of NetworkGraph as it has active changes
- Separate Interpreter to a different header.

myungjoo commented:
Would someone explain the idea to me in person next week? probably April 5th? (do not prepare any materials that are not already there, please)

zhoonit commented:
> Would someone explain the idea to me in person next week? probably April 5th? (do not prepare any materials that are not already there, please)

Yes, I'd like to. I'll visit.

kparichay commented on nntrainer/graph/network_graph.h:
```diff
@@ -23,6 +23,7 @@
 #include <vector>
 
 #include <layer_internal.h>
+#include <loss_layer.h>
```
is this intended?

zhoonit replied:
`int addLossLayer(const LossType loss_type);` requires LossType definition, I think we should move loss type definition to other place later.

kparichay replied:
yes.

zhoonit replied:
Added here to track https://github.com/nnstreamer/nntrainer/issues/1063#issuecomment-813175687

kparichay commented on test/meson.build:
```diff
@@ -19,8 +16,17 @@ nntrainer_testutil_dep = declare_dependency(
+  gtest_dep
+]
+
+nntrainer_test_main_deps = [
```
missing usage

kparichay commented on test/meson.build:
```diff
@@ -19,8 +16,17 @@ nntrainer_testutil_dep = declare_dependency(
+  gtest_dep
+]
+
+nntrainer_test_main_deps = [
```
please add a comment how this is different from `nntrainer_test_deps`.

zhoonit replied:
Added

kparichay commented on nntrainer/compiler/compiler.h:
```diff
@@ -0,0 +1,77 @@
+   * @param file ifstream to serialize graph
+   */
+  virtual std::shared_ptr<ExecutableGraph>
+  compile(std::shared_ptr<const GraphRepresentation> representation) = 0;
```
add a todo to take a delegate parameter in this function or in constructor?

zhoonit replied:
I am thinking to delegate through a interface. This might be done outside of the class :)

kparichay replied:
delegate contains two parts - backend `B` (nntrainer, tflite, nnapi etc) and hardware `H` (cpu, gpu, etc). Do you plan to create  `B*H` interfaces?

zhoonit replied:
I was thinking B*H.  (nntrainerCPU, nntrainerGPU), as they would have different implementation. If considerable amount of codes can be shared, it might be better to have a delegate inside the constructor.

kparichay replied:
Starting with either way works. FYI: with `B*H` interfaces, `delegate` will be resolved in the model, when creating the compiler object.

zhoonit replied:
Done

jijoongmoon replied:
I wonder that if you think nntrainerCPU, nntrainerGPU.. then how do you handle CPU+GPU Case?? or heterogeneous case??

zhoonit replied:
We discussed offline that this is very complicated problem and shouldn't be fixated without comprehended survey. This will remain unfixed and codes will consider that as well.

kparichay reviewed (approved):
LGTM

dongju-chae commented on nntrainer/compiler/meson.build:
```diff
@@ -0,0 +1,13 @@
+compiler_sources = [
+]
```
`[]`?

zhoonit replied:
Done :)

dongju-chae reviewed (approved):
Looks cool

jijoongmoon commented on nntrainer/compiler/compiler.h:
```diff
@@ -0,0 +1,79 @@
+ *      +---------------+
+ *
+ * @see	https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
```
maybe @details goes last?

zhoonit replied:
Yes, I'll make a PR to update this :)

jijoongmoon commented on nntrainer/compiler/compiler.h:
```diff
@@ -0,0 +1,79 @@
+   * @return GraphRepresentation graph representation
+   */
+  virtual std::shared_ptr<GraphRepresentation>
+  decompile(std::shared_ptr<ExecutableGraph> executable) = 0;
```
does this decompile works only for nntrainer? Cause, GraphRepresentation and ExcutableGraph is from nntrainer.

zhoonit replied:
Yes, until we consider more complicated structure.

jijoongmoon commented on nntrainer/compiler/interpreter.h:
```diff
@@ -0,0 +1,79 @@
+ *
+ */
+class GraphInterpreter {
+public:
```
I wonder that if we are going to have nntrainerGraphInterpreter and tfliteGraphInterpreter and others??

zhoonit replied:
Yes, those will be following PR :)

jijoongmoon reviewed (approved):
LGTM for now. We need to improve the design of the compiler as we develop.
